### PR TITLE
doc: interrupts: add documentation section for zero-latency IRQs

### LINKED ISSUE #21185: zero-latency IRQ behavior is not documented?
ARM zero-latency IRQ functionality was added some time this year, but there appears to be no documentation on what it does and its interaction with other thread and interrupt behavior.

From what I can tell, zero-latency IRQs will be taken even when interrupts are disabled, and there is a hint that meta-irq priority threads in the future may be given control from a zero-latency IRQ.

It would help if the behavior of this feature were documented, and suggestions provided about how to mitigate latency induced as a result of such an interrupt, and perhaps restrictions on what is permitted within a ZLI handler to prevent breaking otherwise working code due to broken locks.

### PR DESCRIPTION
**Pull request by ioannisg**
Add a simple documentation section for the Zero-Latency
IRQs feature supported by the kernel.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

Closes #21185

### CODE REVIEW

File: doc/reference/kernel/other/interrupts.rst [hunk 1]
```diff
@@ -152,6 +152,28 @@ The IRQ must be subsequently **enabled** to permit the ISR to execute.
...
+    The kernel addresses such use-cases by allowing interrupts with critical
+    latency constraints to execute at a priority level that cannot be blocked
+    by interrupt locking. These interrupts are defined as
+    *zero-latency interrupts*. The feature is enabled using
+    :option:`CONFIG_ZERO_LATENCY_IRQS`.
+
+    As zero-latency interrupts may preempt the execution of kernel critical
+    section operations they shall not be allowed to use any kernel
+    functionality that may modify kernel structures, or generate exceptions
+    that need to be handler asynchronously (e.g. kernel panic).
```
Comment (pabigot) [inline doc/reference/kernel/other/interrupts.rst]: My understanding is that taking a spinlock on a uniprocessor involves blocking interrupts, and that ZLIs can bypass that, therefore they run while locks are being held.  What I'd like to see, which relates to #21648, is clear description of what you can do in a ZLI, referencing terms from #21678.  "shall not...modify kernel structures" is imprecise. 

Clearly a ZLI must be at least avoid invoking anything that isn't **isr-ok**.

I would expect a ZLI should be able to queue material to be processed by a thread, e.g. invoke `k_sem_give()` or `k_queue_put()`.  So it has to be able to invoke things that are **reschedule**, and so can modify some kernel structures, right?

However, what if there's a meta-IRQ thread that gets made ready by an operation in a ZLI?  At the next reschedule point the meta-IRQ thread will be selected, regardless of scheduler lock and current thread type (e.g. cooperative non-meta-IRQ).

Is there any chance that this could enable a context switch while the spinlock is still held?  Not at the point the ZLI completes, unless that's a reschedule point (is it?  nothing specifies) but later, within the locked code that was interrupted when that code issues `k_sem_give()` which would not normally cause a context switch because it was a non-meta-IRQ cooperative thread?

BTW? Does `k_is_in_isr()` correctly identify that one is in a ZLI?  (As opposed to thread mode; clearly it couldn't distinguish ZLI from non-ZLI ISR.)

Comment (ioannisg) [reply doc/reference/kernel/other/interrupts.rst]: > My understanding is that taking a spinlock on a uniprocessor involves blocking interrupts, and that ZLIs can bypass that, therefore they run while locks are being held. What I'd like to see, which relates to #21648, is clear description of what you can do in a ZLI, referencing terms from #21678. "shall not...modify kernel structures" is imprecise.
> 
> Clearly a ZLI must be at least avoid invoking anything that isn't **isr-ok**.

Yes, but since ZLI is an ISR, this assumption is satisfied already.

> 
> I would expect a ZLI should be able to queue material to be processed by a thread, e.g. invoke `k_sem_give()` or `k_queue_put()`. So it has to be able to invoke things that are **reschedule**, and so can modify some kernel structures, right?

This is not necessary, if we accept that ZLI will be a sub-arch feature. Currently it is used in the Nordic LE Controller and it does not touch the kernel.

> 
> However, what if there's a meta-IRQ thread that gets made ready by an operation in a ZLI? At the next reschedule point the meta-IRQ thread will be selected, regardless of scheduler lock and current thread type (e.g. cooperative non-meta-IRQ).
> 
This is not applicable we don't allow ZLIs to touch kernel.

> Is there any chance that this could enable a context switch while the spinlock is still held? Not at the point the ZLI completes, unless that's a reschedule point (is it? nothing specifies) but later, within the locked code that was interrupted when that code issues `k_sem_give()` which would not normally cause a context switch because it was a non-meta-IRQ cooperative thread?
> 

ZLIs are wrapped around the standard ISR wrapper so there is a reschedule check at the end, anyways. But this should not be an issue if ZLI code does nothing to generate conditions for reschedule, right?

> BTW? Does `k_is_in_isr()` correctly identify that one is in a ZLI? (As opposed to thread mode; clearly it couldn't distinguish ZLI from non-ZLI ISR.)

Correct, it does not distinguish between the two types of ISRs.

File: doc/reference/kernel/other/interrupts.rst [hunk 2]
```diff
@@ -152,6 +152,28 @@ The IRQ must be subsequently **enabled** to permit the ISR to execute.
     Disabling an IRQ prevents *all* threads in the system from being preempted
     by the associated ISR, not just the thread that disabled the IRQ.
 
+Zero Latency Interrupts
+-----------------------
+
+    Preventing interruptions by applying an IRQ lock may increase the observed
```
Comment (nashif) [inline doc/reference/kernel/other/interrupts.rst]: why the indentation here?

File: doc/reference/kernel/other/interrupts.rst (hunk 1 above)
Comment (pabigot) [reply doc/reference/kernel/other/interrupts.rst]: > > I would expect a ZLI should be able to queue material to be processed by a thread, e.g. invoke `k_sem_give()` or `k_queue_put()`. So it has to be able to invoke things that are **reschedule**, and so can modify some kernel structures, right?
> 
> This is not necessary, if we accept that ZLI will be a sub-arch feature. Currently it is used in the Nordic LE Controller and it does not touch the kernel.

OK.  As long as nobody ever wants to use a ZLI to capture and quickly clear an event that requires a thread be notified that the event happened, great.

> > BTW? Does `k_is_in_isr()` correctly identify that one is in a ZLI? (As opposed to thread mode; clearly it couldn't distinguish ZLI from non-ZLI ISR.)
> 
> Correct, it does not distinguish between the two types of ISRs.

I should have been more clear: does `k_is_in_isr()` return true when invoked from a ZLI?  I wasn't aware that ZLIs used the standard ISR wrapper, so I guess the answer is probably "yes".

File: doc/reference/kernel/other/interrupts.rst [hunk 3]
```diff
@@ -152,6 +152,28 @@ The IRQ must be subsequently **enabled** to permit the ISR to execute.
...
+    interrupt latency. A high interrupt latency, however, may not be acceptable
+    for certain low-latency use-cases.
+
+    The kernel addresses such use-cases by allowing interrupts with critical
+    latency constraints to execute at a priority level that cannot be blocked
+    by interrupt locking. These interrupts are defined as
+    *zero-latency interrupts*. The feature is enabled using
+    :option:`CONFIG_ZERO_LATENCY_IRQS`.
```
Comment (pabigot) [inline doc/reference/kernel/other/interrupts.rst]: When people are reading this they might say "OK, so that's how I enable it; what do I do next?".  A link to `ISR_DECLARE_DIRECT()` would answer that question; here,  in the description of the option, or both.

File: doc/reference/kernel/other/interrupts.rst (hunk 1 above)
Comment (andyross) [reply doc/reference/kernel/other/interrupts.rst]: I wouldn't mind a somewhat stronger statement here either.   A ZLI will defeat any synchronization the kernel tries to do.  Effectively the ZLI code lives "outside" the kernel and may not do anything that modifies or even inspects state that the kernel expects to be synchronized.  We'll never be able to document all that rigorously, so IMHO the docs should make the "outside" requirement clear.  Something like:

*"Zero latency interrupts are expected to be used to manage hardware events directly, and not to interoperate with the kernel code at all.  They should treat all kernel APIs as undefined behavior (i.e. an app that uses them in a ZLI is responsible for directly verifying correct behavior), and they should not modify any data inspected by kernel APIs invoked from normal Zephyr contexts."*

File: doc/reference/kernel/other/interrupts.rst (hunk 2 above)
Comment (ioannisg) [reply doc/reference/kernel/other/interrupts.rst]: typo, I suppose. Will fix.

File: doc/reference/kernel/other/interrupts.rst (hunk 3 above)
Comment (ioannisg) [reply doc/reference/kernel/other/interrupts.rst]: Hm, right, I will rephrase this. I only wanted to say that the KConfig option enables the support of the feature.

File: doc/reference/kernel/other/interrupts.rst (hunk 1 above)
Comment (ioannisg) [reply doc/reference/kernel/other/interrupts.rst]: Thanks for the feedback @andyross . Well, I don't see any hard with inspecting kernel structures, as long as no modification is done, but I admit the value of a stronger statement here. I'll update it

Comment (andyross) [reply doc/reference/kernel/other/interrupts.rst]: You can't break the kernel by reading its data, but you can still make wrong decisions.  Consider inspecting things like "Does this semaphore have any waiting threads?" -- you may have interrupted the kernel in the middle of a sempahore operation and the answer may depend on which state you inspect.

Even things like "What time is it?" can break.  The tick count is 64 bit, and you may have interrupted the timer ISR while it had written only half of a 32-bit-rolled-over value.   (FWIW: exactly this issue of non-atomic time updates vs. interrupts was the source of one of the first really hard embedded bugs I ever figured out, on a 8051 system about 21 years ago, heh).

Basically: we should promise that nothing works.  And if something can be made to work, we should clarify that it's the application author's job to do all the validation.

Comment (andrewboie) [reply doc/reference/kernel/other/interrupts.rst]: I concur with @andyross on this.

File: doc/reference/kernel/other/interrupts.rst [hunk 4]
```diff
@@ -152,6 +152,28 @@ The IRQ must be subsequently **enabled** to permit the ISR to execute.
...
+
+    As zero-latency interrupts may preempt the execution of kernel critical
+    section operations they shall not be allowed to use any kernel
+    functionality that may modify kernel structures, or generate exceptions
+    that need to be handler synchronously (e.g. kernel panic).
```
Comment (andrewboie) [inline doc/reference/kernel/other/interrupts.rst]: "handled"

File: doc/reference/kernel/other/interrupts.rst [hunk 5]
```diff
@@ -152,6 +152,28 @@ The IRQ must be subsequently **enabled** to permit the ISR to execute.
...
+
+.. important::
+    Zero-latency interrupts are supported on an architecture-specific basis.
+    The feature is currently implemented in ARM Cortex-M architecture variant.
```
Comment (andrewboie) [inline doc/reference/kernel/other/interrupts.rst]: "in the ARM"

File: doc/reference/kernel/other/interrupts.rst (hunk 2 above)
Comment (ioannisg) [reply doc/reference/kernel/other/interrupts.rst]: done

File: doc/reference/kernel/other/interrupts.rst (hunk 3 above)
Comment (ioannisg) [reply doc/reference/kernel/other/interrupts.rst]: updated

File: doc/reference/kernel/other/interrupts.rst (hunk 4 above)
Comment (ioannisg) [reply doc/reference/kernel/other/interrupts.rst]: thanks

File: doc/reference/kernel/other/interrupts.rst (hunk 5 above)
Comment (ioannisg) [reply doc/reference/kernel/other/interrupts.rst]: sure

File: doc/reference/kernel/other/interrupts.rst (hunk 1 above)
Comment (ioannisg) [reply doc/reference/kernel/other/interrupts.rst]: updated, as agreed, based on the input from @andyross .
I only kept the original phrase about not allowed synchronous exceptions, to stress that a bit more.

**Review by andyross — APPROVED** [review]

**Review by nashif — APPROVED** [review]
